crates/kitsune_p2p/mdns/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+///! Crate for discovering Holochain peers over MDNS
+///! Works by broadcasting a service named `HC_SERVICE_NAME`
+///! and adding base64 encoded data in a TXT record
+///!
+///! Uses libmdns crate for broadcasting
+///! Uses mdns crate for discovery
+
+use futures_util::{stream::StreamExt, self};
+use std::thread;
+use std::time::Duration;
+use std::sync::mpsc::{self, TryRecvError, Sender};
+use mdns::RecordKind;
+use futures_core::Stream;
+use err_derive::Error;
+
+const HC_SERVICE_NAME: &str       = "_holochain._udp";
+const BROADCAST_INTERVAL_SEC: u64 = 8;
+const QUERY_INTERVAL_SEC: u64     = 5;
+
+#[derive(Debug, Error)]
+pub enum MdnsError {
+   #[error(display = "Regular Mdns error {}", _0)]
+   Mdns(#[error(source)] mdns::Error),
+   #[error(display = "Base64 decoding error {}", _0)]
+   Base64(#[error(source)] base64::DecodeError),
+}
+
+/// Stop thread created by `mdns_create_broadcast_thread()`
+pub fn mdns_kill_thread(tx: Sender<()>) {
+   tx.send(()).ok();
+}
+
+/// Create a thread that will broadcast the holochain service over mdns
+/// Returns Sender for sending thread termination command
+pub fn mdns_create_broadcast_thread(buffer: &[u8]) -> Sender<()> {
+   // Create Terminate command channel
+   let (tx, rx) = mpsc::channel();
+   // Change buffer to base64 string
+   let b64 =format!("u{}", base64::encode_config(buffer, base64::URL_SAFE_NO_PAD));
+   println!("base64 length is {}", b64.len());
+
+   // Create thread
+   let _handle = thread::spawn(move || {
+      // debug
+      println!("Entering mdns broadcasting thread...");
+      // Create mdns responder
+      let responder = libmdns::Responder::new().unwrap();
+      let _svc = responder.register(
+         HC_SERVICE_NAME.to_owned(),
+         "holonode".to_owned(),
+         80,
```

Comment:
is `0` valid? Since we won't be using the port from here, setting it to `80` seems a little strange : )